rtos/TARGET_CORTEX/mbed_rtx_idle.cpp

```diff
@@ -95,21 +95,22 @@ uint32_t OS_Tick_GetInterval (void) {
 
 static void default_idle_hook(void)
 {
-    uint32_t elapsed_ticks = 0;
+    uint32_t ticks_to_sleep = osKernelSuspend();
+    os_timer->suspend(ticks_to_sleep);
 
-    core_util_critical_section_enter();
-    uint32_t ticks_to_sleep = svcRtxKernelSuspend();
-    if (ticks_to_sleep) {
-        os_timer->schedule_tick(ticks_to_sleep);
+    while (!os_timer->suspend_time_passed()) {
 
-        sleep();
+        core_util_critical_section_enter();
+        if (!osRtxInfo.kernel.pendSV) { // check devices didn't run any interrupt handlers kicking the OS between suspend and critical section
+            sleep();                    // uses WFI - will wake as soon as any pending interrupts
+        }
+        core_util_critical_section_exit();
```

Comment:
There might be call for an ISB here - you're wanting pending interrupts to trigger between this line and the next one, and an ISB should make sure that happens.

In theory I think you could redisable interrupts before any have a chance to trigger - formally the sequence "enable IRQs / disable IRQs" needs an ISB somewhere in the middle to guarantee they have time to trigger.

Relevant ARM ARM quote:

> If execution of a CPS instruction:
> * Increases the execution priority, the CPS execution serializes that change to the instruction stream.
> * Decreases the execution priority, the architecture guarantees only that the new priority is visible to instructions executed after either executing an ISB, or performing an exception entry or exception return.

(In practice they're likely to trigger before you hit the next disable, which means you'd see pendSV set by the next time you hit line 104.)